src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -607,17 +607,34 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
 #nullable restore
             JsonSerializerOptions options);
 
-        internal virtual T ReadWithQuotes(ref Utf8JsonReader reader)
+        internal virtual T ReadFromPropertyName(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            if (!IsInternalConverter && options.TryGetDefaultSimpleConverter(TypeToConvert, out JsonConverter? defaultConverter))
+            {
+                // .NET 5 backward compatibility: hardcode the default converter for primitive key serialization.
+                Debug.Assert(defaultConverter.IsInternalConverter && defaultConverter is JsonConverter<T>);
+                return ((JsonConverter<T>)defaultConverter).ReadFromPropertyName(ref reader, TypeToConvert, options);
+            }
+
             ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
             return default;
         }
 
-        internal virtual void WriteWithQuotes(Utf8JsonWriter writer, [DisallowNull] T value, JsonSerializerOptions options, ref WriteStack state)
-            => ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
+        internal virtual void WriteToPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options, ref WriteStack state)
```

Comment:
Assuming you wanna remove `ref WriteStack state` in a follow up... It's currently used to honor options.DictionaryKeyPolicy for `string` and `Enum` keys, so we'll need some special-casing in those converters.

Reply:
Yeah I think we need to keep this as an internal overload similar to what we do with the `TryWrite` method.